milli/src/search/criteria/exactness.rs

```diff
@@ -360,3 +329,452 @@ impl ExactQueryPart {
+/// ```text
+/// 3: [b0&b2&b3 | b1&b2&b3]
+/// 4: [b0&b2&b4 | b1&b2&b4]
+/// 5: [b0&b2&b5 | b1&b2&b5]
+/// etc.
+/// ```
+/// We only keep two maps in memory at any one point. As soon as Map J is built, we flatten Map J-1 into
+/// a single bitmap by taking the union of all of its values. This union gives us Xj-1.
+///
+/// ## Memory Usage
+/// This function is expected to be called on a maximum of 10 bitmaps. The worst case thus happens when
+/// 10 identical large bitmaps are given.
+///
+/// In the context of Meilisearch, let's imagine that we are given 10 bitmaps containing all
+/// the document ids. If the dataset contains 16 million documents, then each bitmap will take
+/// around 2MB of memory.
+///
+/// When creating Map 3, we will have, in memory:
+/// 1. The 10 original bitmaps (20MB)
+/// 2. X0 : 2MB
+/// 3. Map 1, containing 9 bitmaps: 18MB
+/// 4. Map 2, containing 8 bitmaps: 16MB
+/// 5. X1: 2MB
+/// for a total of around 60MB of memory. This roughly represents the maximum memory usage of this function.
+///
+/// ## Time complexity
+/// Let N be the size of the given list of bitmaps and M the length of each individual bitmap.
+///
+/// We need to create N new bitmaps. The most expensive one to create is the second one, where we need to
+/// iterate over the N keys of Map 1, and for each of those keys `k_i`, we perform `N-k_i` bitmap unions.
+/// Unioning two bitmaps is O(M), and we need to do it O(N^2) times.
+///
+/// Therefore the time complexity is O(N^3 * M).
+fn create_non_disjoint_combinations(bitmaps: Vec<RoaringBitmap>) -> Vec<RoaringBitmap> {
+    let nbr_parts = bitmaps.len();
+    if nbr_parts == 1 {
+        let flattened_base_level = MultiOps::union(bitmaps.into_iter());
+        return vec![flattened_base_level];
+    }
+    let mut flattened_levels = vec![];
+    let mut last_level: BTreeMap<usize, RoaringBitmap> =
+        bitmaps.clone().into_iter().enumerate().collect();
+
+    for _ in 2..=nbr_parts {
+        let mut new_level = BTreeMap::new();
+        for (last_part_index, base_combination) in last_level.iter() {
+            #[allow(clippy::needless_range_loop)]
+            for new_last_part_index in last_part_index + 1..nbr_parts {
+                let new_combination = base_combination & &bitmaps[new_last_part_index];
+                if !new_combination.is_empty() {
+                    match new_level.entry(new_last_part_index) {
+                        Entry::Occupied(mut b) => {
+                            *b.get_mut() |= new_combination;
+                        }
+                        Entry::Vacant(entry) => {
+                            entry.insert(new_combination);
+                        }
+                    }
+                }
+            }
+        }
+        // Now flatten the last level to save memory
+        let flattened_last_level = MultiOps::union(last_level.values());
+        flattened_levels.push(flattened_last_level);
+        last_level = new_level;
```

Comment:
In this case, you want to consume `last_level` and then reassign it with `new_level`, it may be better to use [into_value](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.into_values):
```suggestion
        let level = std::mem::take(&mut last_level);
        let flattened_last_level = MultiOps::union(level.into_values());
        flattened_levels.push(flattened_last_level);
        last_level = new_level;
```